kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -648,4 +651,47 @@ protected static void validateSupportedType(DataType dataType) {
       throw unsupportedDataType(dataType);
     }
   }
+
+  /**
+   * If IcebergWriterCompatV1 is enabled, we need to ensure that map struct keys don't change. This
+   * validates that
+   */
+  private static void validateNoMapStructKeyChanges(
+      StructType currentSchema, StructType newSchema, Map<String, String> configuration) {
+    if (TableConfig.ICEBERG_WRITER_COMPAT_V1_ENABLED.fromMetadata(configuration)) {
+      // do the check if the feature is enabled now
+      List<Tuple2<List<String>, StructField>> columnPathToMaps =
+          filterRecursively(
+              currentSchema,
+              true, /* recurseIntoMapOrArrayElements, need to check maps inside arrays */
+              false, /* stopOnFirstMatch */
+              sf -> {
+                DataType dataType = sf.getDataType();
+                return (dataType instanceof MapType)
+                    && ((MapType) dataType).getKeyField().getDataType() instanceof StructType;
+              });
+      Map<Integer, StructField> updatedFieldsById = fieldsById(newSchema);
+      for (Tuple2<List<String>, StructField> pathAndField : columnPathToMaps) {
+        StructField newMap = updatedFieldsById.get(getColumnId(pathAndField._2));
+        if (newMap != null) {
+          // if the map was removed, that's okay, so only check if newMap is not null
+          DataType newDataType = newMap.getDataType();
+          if (newDataType instanceof MapType) {
+            // This check doesn't block changing the whole type, so if it's not a Map, don't bother
+            // to check anything
+            StructField currentStructKey = ((MapType) pathAndField._2.getDataType()).getKeyField();
+            // cast is safe, we filtered for this
+            StructType currentKeyType = (StructType) currentStructKey.getDataType();
+            StructField newStructKey = ((MapType) newDataType).getKeyField();
+            if (!currentKeyType.equals(newStructKey.getDataType())) {
```

Comment:
I think this check doesn't allow re-arranging fields (should it?), or updating the column metadata, changing nullability, etc. What's the expected behavior for these kinds of changes?

Also, I don't think this is necessarily the solution, but consider the difference in `StructType::equivalent` vs `StructType::equals`, but maybe neither of those are the comparison needed.

Reply:
Also should you be able to for example, widen the type, of a nested key field?

Reply:
I see in the spec RFC it says Iceberg blocks any change not just additions. If this is true I guess this is okay? If you can confirm all the above listed changes should in fact be blocked this seems fine to me

Reply:
Yes, we're aiming to just block _any_ changes for now. We can relax it in the future if we find it to be necessary. I think the basic idea is that changes could change the hash code and mean you have to recompute everything